src/Grid/grid.jl

```diff
@@ -326,18 +326,18 @@ end
 # Grid utility functions #
 ##########################
 """
-    getneighbors(grid::Grid{dim,C,T,ExclusiveTopology}, cellidx::CellIndex, include_self=false)
-    getneighbors(grid::Grid{dim,C,T,ExclusiveTopology}, faceidx::FaceIndex, include_self=false)
-    getneighbors(grid::Grid{dim,C,T,ExclusiveTopology}, vertexidx::VertexIndex, include_self=false)
-    getneighbors(grid::Grid{dim,C,T,ExclusiveTopology}, EdgeIndex::EdgeIndex, include_self=false)
+    getneighborhood(grid::Grid{dim,C,T,ExclusiveTopology}, cellidx::CellIndex, include_self=false)
+    getneighborhood(grid::Grid{dim,C,T,ExclusiveTopology}, faceidx::FaceIndex, include_self=false)
+    getneighborhood(grid::Grid{dim,C,T,ExclusiveTopology}, vertexidx::VertexIndex, include_self=false)
+    getneighborhood(grid::Grid{dim,C,T,ExclusiveTopology}, EdgeIndex::EdgeIndex, include_self=false)
 
 Returns all directly connected entities of the same type, i.e. calling the function with a `VertexIndex` will return
 a list of directly connected vertices (connected via face/edge). If `include_self` is true, the given `*Index` is included 
```

Comment:
Just as a reminer, with the current interface we cannot directly control which kind of neighborhood we want? E.g.for vertices just the edge or just the face neighborhood.

Reply:
No, but implementation on top of it is possible, e.g. given a vertex get all directly connected edges would require to get all edges of the current cell and the cell neighbors were the given vertex is part of a cell edge, etc.

Reply:
I know that it is possible to do. :) My question aimed at our abstraction layer, because we want to leave the option open to utilize different topologies, which may be optimized for specific use cases. However, we can develop some algorithms on top of a common interface which are independent of the actual topology.

Reply:
but this is still open, if you dispatch simply on a different topology type, instead of `ExclusiveTopology`

Reply:
Right, but currently we have the contract that "a list of directly connected vertices (connected via face/edge)" is returned, which is ambiguous. E.g. a new topology may return the vertices connected via edges on hexahedral elements which another one may return the vertices connected via faces. Both have legitimate use cases, but the contract is too wide imho. Similar issues arise due to ambiguities in NC meshes. Not sure how to improve the docs here.

Reply:
is directly connected really ambigious? With directly connected I mean all vertices connected by a face in 2D or an edge in 3D, i.e. where you have a direct line in between

Reply:
from the example from my comment below about NC meshes, e.g. node 7
```
julia> grid =  Grid(cells,nodes;topology=top)
Grid{2, Quadrilateral, Int64, Ferrite.ExclusiveTopology} with 3 Quadrilateral cells and 8 nodes

julia> Ferrite.getneighborhood(grid,VertexIndex(1,3))
4-element Vector{VertexIndex}:
 VertexIndex((1, 2))
 VertexIndex((1, 4))
 VertexIndex((3, 1))
 VertexIndex((3, 3))
```

Reply:
Yes, I think it is ambiguous. 
1. we have no definition of what a face/edge is (if you remember the discussion on slack). 
2. currently we use face and edge in 2D to **sometimes** refer to the same entity and sometimes not (due to the missing definitions)
3. while I intuitively understand what you want to say with "connected by a direct line", you can also draw a line over faces to connect the "diagonal" in hex faces. Not sure how to precise this one.

The issue is clearer for the nodes 2 and 7. Is 5 really a neighbor for them? And how relates 3 to them? However, I would leave the NC issues out of the PR for now to not overcomplicate it.

Reply:
I don't see how this results in something ambiguous in this particular example if the definitions of `faces, edges` and `vertices`is used from: https://github.com/Ferrite-FEM/Ferrite.jl/blob/master/src/Grid/grid.jl#L459-L476
in 2D the line is the face, while in 3D the line is the edge of a three dimensional cell. The face of a two dimensional connected to a 3D cell results in a face/edge neighborhood in my code, while a truly embedded 2D cell in 3D has again a edge neighborhood https://github.com/Ferrite-FEM/Ferrite.jl/blob/master/src/Grid/grid.jl#L475. Such that we end up with something that can distinguish between the "line entities" of a given cell, due such if clauses https://github.com/Ferrite-FEM/Ferrite.jl/blob/mk/topologyinfo/src/Grid/grid.jl#L209 in combination with the dispatching of the embedded cells in e.g. https://github.com/Ferrite-FEM/Ferrite.jl/blob/mk/topologyinfo/src/Grid/grid.jl#L718 and https://github.com/Ferrite-FEM/Ferrite.jl/blob/mk/topologyinfo/src/Grid/grid.jl#L724

So truly I'm relying on the given definitions. If it's a problem of the definitions, then it's not the related to this PR. However the PR or the implementation *should* rely always on these definitions, since they allow for a user defined cell extension. 

To finalize it, the situation is very clear (at least for me) for node 5, 5 is a vertex neighbor of those nodes, why not? It is directly connected by the corresponding line entity of the cell, i.e. face
```
julia> nodes = [Ferrite.Node((0,0)) for i in 1:8];

julia> cells = [Quadrilateral((4,5,7,6)),Quadrilateral((1,2,5,4)),Quadrilateral((2,3,8,7))];

julia> top = Ferrite.ExclusiveTopology(cells);

julia> grid =  Grid(cells,nodes;topology=top);

julia> Ferrite.getneighborhood(grid,VertexIndex(3,4))
4-element Vector{VertexIndex}:
 VertexIndex((1, 2))
 VertexIndex((1, 4))
 VertexIndex((3, 1))
 VertexIndex((3, 3))

julia> Ferrite.getneighborhood(grid,VertexIndex(1,3))
4-element Vector{VertexIndex}:
 VertexIndex((1, 2))
 VertexIndex((1, 4))
 VertexIndex((3, 1))
 VertexIndex((3, 3))

julia> Ferrite.getneighborhood(grid,VertexIndex(2,2))
4-element Vector{VertexIndex}:
 VertexIndex((2, 1))
 VertexIndex((2, 3))
 VertexIndex((3, 2))
 VertexIndex((3, 4))

julia> Ferrite.getneighborhood(grid,VertexIndex(3,1))
4-element Vector{VertexIndex}:
 VertexIndex((2, 1))
 VertexIndex((2, 3))
 VertexIndex((3, 2))
 VertexIndex((3, 4))

```

Reply:
> [...] If it's a problem of the definitions, then it's not the related to this PR. [...]

Agreed. Since this is quite complicated let us leave the disambiguation issue for now. Maybe we should leave a comment on the strict dependency on the mentioned interface to vertices,edges and faces in the topology docs?

> To finalize it, the situation is very clear (at least for me) for node 5, 5 is a vertex neighbor of those nodes, why not? [...]

Okay, I was not precise enough in the problem description here, so let me leave a clarification for future reference. Depending on how you define "face" and "line" in NC meshes along the "nonconforming face", you can have the following neighborhoods for vertex 3:
* {1,3,5,7}
* {1,3,7}
* {1,3,5}

Maybe, after the last design decisions for this PR are settled, we should move some parts of your examples to the docs (since I think they are quite nice for new users to understand the interface).